client/fingerprint/dynamic_host_volumes.go

```diff
@@ -0,0 +1,119 @@
+	hvm "github.com/hashicorp/nomad/client/hostvolumemanager"
+	"github.com/hashicorp/nomad/helper"
+)
+
+func NewPluginsHostVolumeFingerprint(logger hclog.Logger) Fingerprint {
+	return &DynamicHostVolumePluginFingerprint{
+		logger: logger.Named("host_volume_plugins"),
+	}
+}
+
+var _ ReloadableFingerprint = &DynamicHostVolumePluginFingerprint{}
+
+type DynamicHostVolumePluginFingerprint struct {
+	logger hclog.Logger
+}
+
+func (h *DynamicHostVolumePluginFingerprint) Reload() {
+	// host volume plugins are re-detected on agent reload
+}
+
+func (h *DynamicHostVolumePluginFingerprint) Fingerprint(request *FingerprintRequest, response *FingerprintResponse) error {
+	// always add "mkdir" plugin
+	h.logger.Debug("detected plugin built-in", "plugin_id", "mkdir", "version", hvm.HostVolumePluginMkdirVersion)
+	defer response.AddAttribute("plugins.dhv.version.mkdir", hvm.HostVolumePluginMkdirVersion)
+	response.Detected = true
+
+	// this config value will be empty in -dev mode
+	pluginDir := request.Config.DynamicHostVolumePluginPath
+	if pluginDir == "" {
+		return nil
+	}
+
+	plugins, err := GetHostVolumePluginVersions(h.logger, pluginDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			h.logger.Debug("plugin dir does not exist", "dir", pluginDir)
+		} else {
+			h.logger.Warn("error finding plugins", "dir", pluginDir, "error", err)
+		}
+		return nil // don't halt agent start
+	}
+
+	// if this was a reload, wipe what was there before
+	for k := range request.Node.Attributes {
+		if strings.HasPrefix(k, "plugins.dhv.") {
+			response.RemoveAttribute(k)
+		}
+	}
+
+	// set the attribute(s)
+	for plugin, version := range plugins {
+		h.logger.Debug("detected plugin", "plugin_id", plugin, "version", version)
+		response.AddAttribute("plugins.dhv.version."+plugin, version)
+	}
+
+	return nil
+}
+
+func (h *DynamicHostVolumePluginFingerprint) Periodic() (bool, time.Duration) {
+	return false, 0 // not periodic;  db TODO(1.10.0): could be?
+}
+
+// GetHostVolumePluginVersions finds all the executable files on disk
+// that respond to a Version call (arg $1 = 'version' / env $OPERATION = 'version')
+// The return map's keys are plugin IDs, and the values are version strings.
+func GetHostVolumePluginVersions(log hclog.Logger, pluginDir string) (map[string]string, error) {
+	files, err := helper.FindExecutableFiles(pluginDir)
+	if err != nil {
+		return nil, err
+	}
+
+	plugins := make(map[string]string)
+	mut := sync.Mutex{}
+	var wg sync.WaitGroup
+
+	for file, fullPath := range files {
+		wg.Add(1)
+		go func(file, fullPath string) {
+			defer wg.Done()
+			// really should take way less than a second
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+			defer cancel()
+
+			log := log.With("plugin_id", file) // db TODO: "name" like CNI, instead of "ID" ?
+
+			p, err := hvm.NewHostVolumePluginExternal(log, file, fullPath, "")
+			if err != nil {
+				log.Debug("error getting plugin", "error", err)
```

Comment:
We might want to be noisier than debug (probably warn), because something has gone fairly wrong in their configuration to hit any of the error cases we could get  from this method. You'd just end up with a missing attribute and no explanation except in the debug logs.

Reply:
:+1: this may be left over from an earlier iteration that would have had duplicate warn/error logs...